### ISSUE TITLE
Add method to detect periodic orbit / "steady-state"

### ISSUE THREAD

**Issue by MichaelClerx** [MichaelClerx/myokit#390]
One idea would be to do a logged simulation, but set the logging interval equal to the cycle length. Ideally this would give nice asymptotic curves that we can plot for debugging.

(Jumping to the predicted states doesn't seem to work in all models, for whatever reason)

**Comment by MichaelClerx** [comment]
![ohara-steady](https://user-images.githubusercontent.com/517644/64100406-703ccd00-cd63-11e9-8efe-f038fb703508.png)

**Comment by MichaelClerx** [comment]
presumably would need an abs as well as a rel tol

**Comment by MichaelClerx** [comment]
Initial version (that doesn't test for alternans)

```
def limit_cycle(model, protocol, rel_tol=1e-5):
    s = myokit.Simulation(model, protocol)
    s.set_tolerance(1e-8, 1e-8)
    cl = protocol.characteristic_time()

    # Get scale of each state
    states = list(model.states())
    d = s.run(cl, log=myokit.LOG_STATE)
    x = np.array([d[var] for var in states])
    scale = np.max(x, axis=1) - np.min(x, axis=1)
    scale[scale==0] = 1

    if debug:
        ts = []
        xs = []
        dxs = []

    ok_count = 0
    max_ok = 3
    beats_per_iteration = 2

    for i in range(5000):
        d = s.run(
            beats_per_iteration * cl,
            log_interval=cl,
            log=myokit.LOG_BOUND + myokit.LOG_STATE,
        )
        x = np.array([d[var] for var in states]).T
        dx = np.abs(x[-1] - x[-2]) / scale

        if np.max(dx) < rel_tol:
            ok_count += 1
        else:
            ok_count = 0

        if debug:
            print(np.max(dx))

            ts.append(d.time())
            xs.append(x)
            dxs.append(np.max(dx))

        if ok_count == max_ok:
            print(
                'Terminating after ' + str(beats_per_iteration * i) + ' beats')
            break

    if ok_count < max_ok:
        print('Terminating after maximum number of beats.')

    if debug:
        t = np.concatenate(ts)
        x = np.concatenate(xs)
        x /= np.max(x, axis=0)
        dx = np.array(dxs)

        plt.figure()
        plt.subplot(3, 1, 1)
        plt.plot(t, x)
        plt.subplot(3, 1, 2)
        plt.plot(1 + np.arange(len(dx)), dx)
        plt.subplot(3, 1, 3)
        plt.plot(1 + np.arange(len(dx)), np.log(dx))

    return s.state()
```

Not sure where something like this would live

**Comment by MichaelClerx** [comment]
`x = np.array([d[var] for var in states]).T` --> this is a bit silly, could also just use `s.state()`?

**Comment by MichaelClerx** [comment]
Easy alternans test would just be to run 10 paces per iteration, check them all against the first state

**Comment by MichaelClerx** [comment]
Untested version that checks for alternans

def limit_cycle(model, protocol, cl=None, rel_tol=1e-5, max_beats=4000,
                    max_period=10, path=None):
        """
        Pre-paces a model to periodic orbit ("steady state").

        Arguments
        ``model``
        ``protocol``
        ``cl``
        ``rel_tol``
        ``max_beats``
        ``max_period``
        ``path``
        """

        # Create simulation
        s = myokit.Simulation(model, protocol)
        s.set_tolerance(1e-8, 1e-8)
        if cl is None:
            cl = protocol.characteristic_time()

        # Load steady-state from file, if given
        try:
            loaded = myokit.load_state(path)
            s.set_state(loaded)
        except Exception:
            loaded = None
            pass
        else:
            print('Loaded state from ' + str(path))

        # Get scale of each state
        states = list(model.states())
        d = s.run(2 * cl, log=myokit.LOG_STATE)
        x = np.array([d[var] for var in states])
        scale = np.max(x, axis=1) - np.min(x, axis=1)
        scale[scale==0] = 1

        # Check if already at steady-state
        x = np.array([d[var] for var in states]).T
        dx = np.abs(x[0] - x[1]) / scale
        if np.max(dx) < rel_tol:
            return s.state() if loaded is None else loaded

        beats = 0
        period = 0
        duration = max_period * cl
        while beats < max_beats:

            # Run and capture a number of beats
            d = s.run(duration, log_interval=cl, log=myokit.LOG_STATE)
            beats += max_period
            x = np.array([d[var] for var in states]).T

            # Check to see if any of the beats are a repeat of the first beat
            period = 0
            for i in range(1, max_period):
                dx = np.abs(x[0] - x[i]) / scale
                if np.max(dx) < rel_tol:
                    period = i
                    break

            # Repeat found! Check if there's a second repeat
            if period > 0:
                # Simulate more beats if needed
                if 2 * period > max_period:
                    d = s.run(duration, log_interval=cl, log=myokit.LOG_STATE)
                    beats += max_period
                    x = np.concatenate((x, np.array([d[var] for var in states]).T))

                dx = np.abs(x[period] - x[2 * period]) / scale
                if np.max(dx) < rel_tol:
                    print('Terminating after ' + str(beats) + ' beats')
                    break
                else:
                    period = 0

        # Save state to file
        if path is not None:
            print('Saving final state to ' + str(path))
            myokit.save_state(path, s.state())

        if period > 1:
            print('WARNING: Detected alternans with period ' + str(period) + '.')
        elif period == 0:
            print('WARNING: Terminating after maximum number of beats.')
            print('Final dx: ' + str(np.max(dx)))

        return s.state()

**Comment by MichaelClerx** [comment]
Needs a better stopping criterium: Change should be small _and_ decreasing. If it's small but stays the same for several beats, the model does not converge and an error should be raised

**Comment by MichaelClerx** [comment]
Chaste version https://github.com/Chaste/Chaste/blob/release/heart/src/odes/single_cell/SteadyStateRunner.cpp

**Comment by mirams** [comment]
Yep, might be nice for @joeyshuttleworth to put his new code for detecting steady states whilst accounting for numerical noise into both Myokit and Chaste!?

**Comment by MichaelClerx** [comment]
I'd like that! Happy to help too

**Comment by MichaelClerx** [comment]
If going into Myokit, would have to live somewhere inside myokit.lib https://github.com/myokit/myokit/issues/1016
